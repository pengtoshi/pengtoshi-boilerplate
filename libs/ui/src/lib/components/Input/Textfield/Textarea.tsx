import clsx from "clsx";
import { forwardRef, useEffect, useId, useState } from "react";
import type { UIProps } from "../../../props";

export interface TextareaProps extends UIProps.TextArea {
  id?: string;
  label?: string;
  placeholder?: string;
  value?: string | number;
  defaultValue?: string | number;
  guide?: string;
  error?: string;
}

const fieldCommon =
  "flex h-[100px] px-4 py-2.5 gap-2.5 items-center justify-between self-stretch rounded-md border border-gray-300 autofill-hide transition-all duration-300";
const inputCommon =
  "text-14/body peer w-full self-stretch overflow-ellipsis bg-transparent focus:outline-none resize-none";
const fieldColor = "bg-gray-100 hover:bg-gray-50 focus-within:border-primary-500 focus-within:bg-gray-50";
const textColor = "text-gray-950 placeholder:text-gray-400 disabled:text-gray-350";

export const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(function Textarea(
  {
    id,
    label,
    placeholder,
    value,
    defaultValue,
    guide,
    error,
    required,
    disabled,
    maxLength,
    onChange,
    className,
    ...props
  },
  ref,
) {
  const localId = `ui-${useId()}`;
  const inputId = id || localId;

  const [internalValue, setInternalValue] = useState(value ?? defaultValue ?? "");
  const [currentLength, setCurrentLength] = useState(0);

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setInternalValue(e.target.value);
    setCurrentLength(e.target.value.length);
    onChange?.(e);
  };

  useEffect(() => {
    if (value === undefined) return;
    setInternalValue(value);
  }, [value]);

  return (
    <div className={clsx("flex w-full flex-col items-start gap-1.5", className)}>
      <div className={clsx(fieldCommon, fieldColor, disabled && "!bg-gray-200")}>
        <textarea
          id={inputId}
          ref={ref}
          value={internalValue}
          disabled={disabled}
          placeholder={placeholder}
          required={required}
          maxLength={maxLength}
          onChange={handleChange}
          className={clsx(inputCommon, textColor)}
          {...props}
        />
      </div>
      <div className="flex w-full justify-between gap-4">
        <div>
          {!!guide && <div className="text-12/body text-gray-600">{guide}</div>}
          {!!error && <div className="text-12/body text-etc-negative">{error}</div>}
        </div>
        {!!maxLength && (
          <div className="text-12/body text-gray-600">
            {currentLength}/{maxLength}
          </div>
        )}
      </div>
    </div>
  );
});
