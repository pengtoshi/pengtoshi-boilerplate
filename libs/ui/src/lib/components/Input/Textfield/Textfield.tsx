import clsx from "clsx";
import React, { forwardRef, useEffect, useId, useState } from "react";
import type { UIProps } from "../../../props";

export interface TextfieldProps extends UIProps.Input {
  id?: string;
  type?: "number" | "text";
  label?: string;
  leadingIcon?: React.ReactNode;
  trailingIcon?: React.ReactNode;
  placeholder?: string;
  value?: string | number;
  defaultValue?: string | number;
  guide?: string;
  error?: string;
  onClear?: () => void;
}

const getDefaultValue = (type?: "number" | "text", value?: string | number, defaultValue?: string | number) => {
  if (value !== undefined) return value;
  if (defaultValue !== undefined) return defaultValue;
  if (type === "number") return 0;
  return "";
};

const fieldCommon =
  "flex h-10 px-4 gap-2.5 items-center justify-between self-stretch rounded-md border border-gray-300 autofill-hide transition-all duration-300";
const inputCommon = "text-14/body peer w-full self-stretch overflow-ellipsis bg-transparent focus:outline-none";
const fieldColor = "bg-gray-100 hover:bg-gray-50 focus-within:border-primary-500 focus-within:bg-gray-50";
const textColor = "text-gray-950 placeholder:text-gray-400 disabled:text-gray-350";

export const Textfield = forwardRef<HTMLInputElement, TextfieldProps>(function Textfield(
  {
    id,
    type = "text",
    label,
    leadingIcon,
    trailingIcon,
    placeholder,
    value,
    defaultValue,
    guide,
    error,
    required,
    disabled = false,
    onChange,
    className,
    ...props
  },
  ref,
) {
  const localId = `ui-${useId()}`;
  const inputId = id || localId;

  const [internalValue, setInternalValue] = useState(getDefaultValue(type, value, defaultValue));

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInternalValue(e.target.value);
    onChange?.(e);
  };

  useEffect(() => {
    if (value === undefined) return;
    setInternalValue(value);
  }, [value]);

  return (
    <div className={clsx("flex w-full flex-col items-start gap-1.5", className)}>
      <div className={clsx(fieldCommon, fieldColor, disabled && "!bg-gray-200")}>
        {!!leadingIcon && leadingIcon}
        <input
          id={inputId}
          ref={ref}
          disabled={disabled}
          placeholder={placeholder}
          required={required}
          type={type}
          value={internalValue}
          onChange={handleChange}
          className={clsx(inputCommon, textColor)}
          {...props}
        />
        {!!trailingIcon && trailingIcon}
      </div>
      {!!guide && <div className="text-12/body text-gray-600">{guide}</div>}
      {!!error && <div className="text-12/body text-etc-negative">{error}</div>}
    </div>
  );
});
