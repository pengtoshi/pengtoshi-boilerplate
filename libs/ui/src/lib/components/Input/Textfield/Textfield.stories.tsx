import type { StoryFn } from "@storybook/react";
import { useState } from "react";
import type { ComponentProps } from "react";
import { Textfield } from "./Textfield";

export default { component: Textfield };

const Template: StoryFn<ComponentProps<typeof Textfield>> = (args: ComponentProps<typeof Textfield>) => {
  const [value, setValue] = useState("");
  return <Textfield {...args} value={value} onChange={(e) => setValue(e.target.value)} />;
};

export const Default = Template.bind({});
Default.args = {
  className: "w-[240px]",
  placeholder: "Enter text here",
};
