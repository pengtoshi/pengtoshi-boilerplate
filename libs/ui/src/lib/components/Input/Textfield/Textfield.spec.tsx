import { composeStories } from "@storybook/react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as stories from "./Textfield.stories";

describe("TextField", () => {
  Object.entries(composeStories(stories)).forEach(([, Story]) => {
    it("should render", () => {
      const { baseElement } = render(<Story />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
