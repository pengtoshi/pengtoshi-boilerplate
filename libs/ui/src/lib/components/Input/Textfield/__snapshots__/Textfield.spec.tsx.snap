// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render 1`] = `
<body>
  <div>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
    <div
      class="flex w-full flex-col items-start gap-1.5 w-[240px]"
    >
      <div
        class="flex h-10 px-4 gap-2.5 items-center justify-between self-stretch rounded-md border border-gray-300 autofill-hide transition-all duration-300 bg-gray-100 hover:bg-gray-50 focus-within:border-primary-500 focus-within:bg-gray-50"
      >
        <input
          class="text-14/body peer w-full self-stretch overflow-ellipsis bg-transparent focus:outline-none text-gray-950 placeholder:text-gray-400 disabled:text-gray-350"
          id="ui-:r0:"
          placeholder="Enter text here"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex w-full flex-col items-start gap-0.5"
      />
    </div>
  </div>
</body>
`;
